clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc

```diff
@@ -4771,6 +4939,82 @@ sycl::event engine_ext::async_rnn_backward(
   return e;
 }
 // DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|dropout_get_state_size|dpct::dnnl
+// DPCT_DEPENDENCY_BEGIN
+// DnnlUtils|dropout_get_state_size_decl
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+inline
+size_t engine_ext::dropout_get_state_size(){
+  auto rng_engine = oneapi::mkl::rng::philox4x32x10(*_q, 0ull);
+  return oneapi::mkl::rng::get_state_size(rng_engine);
+}
+// DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|async_dropout_forward|dpct::dnnl
+// DPCT_DEPENDENCY_BEGIN
+// DnnlUtils|async_dropout_forward_decl
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+inline
+sycl::event engine_ext::async_dropout_forward(const dropout_desc &desc,
+                                              const memory_desc_ext &src_desc,
+                                              void *src,
+                                              const memory_desc_ext &dst_desc,
+                                              void *dst, void *workspace,
+                                              size_t workspace_size) {
+  if (workspace_size < src_desc.get_size()) {
+    throw std::runtime_error("async_dropout_forward: no sufficient workspace.");
+  }
+  float p = desc.get_probability();
+  if (p == 1.f) {
+    return _q->memset(dst, 0, dst_desc.get_size());
+  } else if (p == 0.f) {
+    return async_reorder(1.f, src_desc, src, 0.f, dst_desc, dst);
+  }
+
+  float scale_factor = 1.f / (1.f - p);
+  void *cache = workspace;
+
+  memory_desc_ext rng_data_desc(
+      ::dnnl::memory::desc(src_desc.get_dims(), ::dnnl::memory::data_type::s32,
+                           src_desc.get_strides()));
+  if (src_desc.get_desc().get_data_type() != ::dnnl::memory::data_type::s32) {
+    cache = allocate(rng_data_desc);
+  }
+  oneapi::mkl::rng::generate(*desc.get_distr(), *desc.get_rng_engine(),
```

Comment:
How about using dropout_desc::generate for these two functions? And remove dropout_desc::get_rng_engine, dropout_desc::get_distr and dropout_desc::get_state.